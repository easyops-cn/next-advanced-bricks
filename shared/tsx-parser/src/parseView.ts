import { parse, type ParseResult as BabelParseResult } from "@babel/parser";
import * as t from "@babel/types";
import type { ParseOptions } from "./interfaces.js";
import { parseLegacyModule } from "./modules/parseLegacyModule.js";
import type { ParsedApp } from "./modules/interfaces.js";
import { parseFile } from "./modules/parseFile.js";

export function parseView(source: string, options?: ParseOptions): ParsedApp {
  const app: ParsedApp = {
    appType: "view",
    modules: new Map(),
    files: [
      {
        filePath: "View.tsx",
        content: source,
      },
    ],
    errors: [],
  };

  let ast: BabelParseResult<t.File> | undefined;
  try {
    ast = parse(source, {
      plugins: ["jsx", "typescript"],
      sourceType: "module",
      errorRecovery: true,
    });
  } catch (error) {
    app.errors.push({
      message: `Failed to parse TSX: ${error}`,
      node: null,
      severity: "fatal",
    });
    return app;
  }

  for (const stmt of ast.program.body) {
    if (t.isExportDefaultDeclaration(stmt)) {
      if (t.isFunctionDeclaration(stmt.declaration)) {
        parseFile("View.tsx", app, ast, options);
        return app;
      }
      break;
    }
  }

  parseLegacyModule("View.tsx", app, ast, options);
  return app;
}
