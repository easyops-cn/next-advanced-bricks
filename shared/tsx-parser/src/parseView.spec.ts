import { readFile } from "node:fs/promises";
import path from "node:path";
import { parseView } from "./parseView.js";

describe("parseView", () => {
  test("should work", async () => {
    const code = await readFile(
      path.join(__dirname, "./__fixtures__/logs.txt"),
      "utf-8"
    );

    const view = parseView(code);
    const simplifiedView = {
      ...view,
      files: undefined,
      entry: {
        ...view.entry!,
        source: undefined,
      },
      modules: [],
    };
    expect(simplifiedView).toMatchSnapshot();
  });
});
