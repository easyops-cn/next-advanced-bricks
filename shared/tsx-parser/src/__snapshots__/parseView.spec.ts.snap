// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`parseView should work 1`] = `
{
  "appType": "view",
  "entry": {
    "contracts": Set {
      "<% \`logs/\${CTX.logId}\` %>",
    },
    "defaultExport": {
      "component": {
        "bindingMap": Map {
          Node {
            "end": 766,
            "loc": SourceLocation {
              "end": Position {
                "column": 13,
                "index": 766,
                "line": 44,
              },
              "filename": undefined,
              "identifierName": "query",
              "start": Position {
                "column": 8,
                "index": 761,
                "line": 44,
              },
            },
            "name": "query",
            "start": 761,
            "type": "Identifier",
          } => {
            "id": Node {
              "end": 766,
              "loc": SourceLocation {
                "end": Position {
                  "column": 13,
                  "index": 766,
                  "line": 44,
                },
                "filename": undefined,
                "identifierName": "query",
                "start": Position {
                  "column": 8,
                  "index": 761,
                  "line": 44,
                },
              },
              "name": "query",
              "start": 761,
              "type": "Identifier",
            },
            "kind": "query",
          },
          Node {
            "end": 795,
            "loc": SourceLocation {
              "end": Position {
                "column": 14,
                "index": 795,
                "line": 45,
              },
              "filename": undefined,
              "identifierName": "logId",
              "start": Position {
                "column": 9,
                "index": 790,
                "line": 45,
              },
            },
            "name": "logId",
            "start": 790,
            "type": "Identifier",
          } => {
            "id": Node {
              "end": 795,
              "loc": SourceLocation {
                "end": Position {
                  "column": 14,
                  "index": 795,
                  "line": 45,
                },
                "filename": undefined,
                "identifierName": "logId",
                "start": Position {
                  "column": 9,
                  "index": 790,
                  "line": 45,
                },
              },
              "name": "logId",
              "start": 790,
              "type": "Identifier",
            },
            "initialValue": "<% QUERY.logId %>",
            "kind": "state",
          },
          Node {
            "end": 805,
            "loc": SourceLocation {
              "end": Position {
                "column": 24,
                "index": 805,
                "line": 45,
              },
              "filename": undefined,
              "identifierName": "setLogId",
              "start": Position {
                "column": 16,
                "index": 797,
                "line": 45,
              },
            },
            "name": "setLogId",
            "start": 797,
            "type": "Identifier",
          } => {
            "id": Node {
              "end": 795,
              "loc": SourceLocation {
                "end": Position {
                  "column": 14,
                  "index": 795,
                  "line": 45,
                },
                "filename": undefined,
                "identifierName": "logId",
                "start": Position {
                  "column": 9,
                  "index": 790,
                  "line": 45,
                },
              },
              "name": "logId",
              "start": 790,
              "type": "Identifier",
            },
            "kind": "setState",
          },
          Node {
            "end": 872,
            "loc": SourceLocation {
              "end": Position {
                "column": 20,
                "index": 872,
                "line": 46,
              },
              "filename": undefined,
              "identifierName": "allExpanded",
              "start": Position {
                "column": 9,
                "index": 861,
                "line": 46,
              },
            },
            "name": "allExpanded",
            "start": 861,
            "type": "Identifier",
          } => {
            "id": Node {
              "end": 872,
              "loc": SourceLocation {
                "end": Position {
                  "column": 20,
                  "index": 872,
                  "line": 46,
                },
                "filename": undefined,
                "identifierName": "allExpanded",
                "start": Position {
                  "column": 9,
                  "index": 861,
                  "line": 46,
                },
              },
              "name": "allExpanded",
              "start": 861,
              "type": "Identifier",
            },
            "initialValue": false,
            "kind": "state",
          },
          Node {
            "end": 888,
            "loc": SourceLocation {
              "end": Position {
                "column": 36,
                "index": 888,
                "line": 46,
              },
              "filename": undefined,
              "identifierName": "setAllExpanded",
              "start": Position {
                "column": 22,
                "index": 874,
                "line": 46,
              },
            },
            "name": "setAllExpanded",
            "start": 874,
            "type": "Identifier",
          } => {
            "id": Node {
              "end": 872,
              "loc": SourceLocation {
                "end": Position {
                  "column": 20,
                  "index": 872,
                  "line": 46,
                },
                "filename": undefined,
                "identifierName": "allExpanded",
                "start": Position {
                  "column": 9,
                  "index": 861,
                  "line": 46,
                },
              },
              "name": "allExpanded",
              "start": 861,
              "type": "Identifier",
            },
            "kind": "setState",
          },
          Node {
            "end": 929,
            "loc": SourceLocation {
              "end": Position {
                "column": 20,
                "index": 929,
                "line": 47,
              },
              "filename": undefined,
              "identifierName": "expandedSet",
              "start": Position {
                "column": 9,
                "index": 918,
                "line": 47,
              },
            },
            "name": "expandedSet",
            "start": 918,
            "type": "Identifier",
          } => {
            "id": Node {
              "end": 929,
              "loc": SourceLocation {
                "end": Position {
                  "column": 20,
                  "index": 929,
                  "line": 47,
                },
                "filename": undefined,
                "identifierName": "expandedSet",
                "start": Position {
                  "column": 9,
                  "index": 918,
                  "line": 47,
                },
              },
              "name": "expandedSet",
              "start": 918,
              "type": "Identifier",
            },
            "initialValue": "<% new Set() %>",
            "kind": "state",
          },
          Node {
            "end": 945,
            "loc": SourceLocation {
              "end": Position {
                "column": 36,
                "index": 945,
                "line": 47,
              },
              "filename": undefined,
              "identifierName": "setExpandedSet",
              "start": Position {
                "column": 22,
                "index": 931,
                "line": 47,
              },
            },
            "name": "setExpandedSet",
            "start": 931,
            "type": "Identifier",
          } => {
            "id": Node {
              "end": 929,
              "loc": SourceLocation {
                "end": Position {
                  "column": 20,
                  "index": 929,
                  "line": 47,
                },
                "filename": undefined,
                "identifierName": "expandedSet",
                "start": Position {
                  "column": 9,
                  "index": 918,
                  "line": 47,
                },
              },
              "name": "expandedSet",
              "start": 918,
              "type": "Identifier",
            },
            "kind": "setState",
          },
          Node {
            "end": 994,
            "loc": SourceLocation {
              "end": Position {
                "column": 16,
                "index": 994,
                "line": 48,
              },
              "filename": undefined,
              "identifierName": "modalRef",
              "start": Position {
                "column": 8,
                "index": 986,
                "line": 48,
              },
            },
            "name": "modalRef",
            "start": 986,
            "type": "Identifier",
          } => {
            "id": Node {
              "end": 994,
              "loc": SourceLocation {
                "end": Position {
                  "column": 16,
                  "index": 994,
                  "line": 48,
                },
                "filename": undefined,
                "identifierName": "modalRef",
                "start": Position {
                  "column": 8,
                  "index": 986,
                  "line": 48,
                },
              },
              "name": "modalRef",
              "start": 986,
              "type": "Identifier",
            },
            "kind": "ref",
          },
          Node {
            "end": 1039,
            "loc": SourceLocation {
              "end": Position {
                "column": 18,
                "index": 1039,
                "line": 49,
              },
              "filename": undefined,
              "identifierName": "detailLog",
              "start": Position {
                "column": 9,
                "index": 1030,
                "line": 49,
              },
            },
            "name": "detailLog",
            "start": 1030,
            "type": "Identifier",
          } => {
            "id": Node {
              "end": 1039,
              "loc": SourceLocation {
                "end": Position {
                  "column": 18,
                  "index": 1039,
                  "line": 49,
                },
                "filename": undefined,
                "identifierName": "detailLog",
                "start": Position {
                  "column": 9,
                  "index": 1030,
                  "line": 49,
                },
              },
              "name": "detailLog",
              "start": 1030,
              "type": "Identifier",
            },
            "initialValue": null,
            "kind": "state",
          },
          Node {
            "end": 1053,
            "loc": SourceLocation {
              "end": Position {
                "column": 32,
                "index": 1053,
                "line": 49,
              },
              "filename": undefined,
              "identifierName": "setDetailLog",
              "start": Position {
                "column": 20,
                "index": 1041,
                "line": 49,
              },
            },
            "name": "setDetailLog",
            "start": 1041,
            "type": "Identifier",
          } => {
            "id": Node {
              "end": 1039,
              "loc": SourceLocation {
                "end": Position {
                  "column": 18,
                  "index": 1039,
                  "line": 49,
                },
                "filename": undefined,
                "identifierName": "detailLog",
                "start": Position {
                  "column": 9,
                  "index": 1030,
                  "line": 49,
                },
              },
              "name": "detailLog",
              "start": 1030,
              "type": "Identifier",
            },
            "kind": "setState",
          },
          Node {
            "end": 1099,
            "loc": SourceLocation {
              "end": Position {
                "column": 13,
                "index": 1099,
                "line": 51,
              },
              "filename": undefined,
              "identifierName": "logs",
              "start": Position {
                "column": 9,
                "index": 1095,
                "line": 51,
              },
            },
            "name": "logs",
            "start": 1095,
            "type": "Identifier",
          } => {
            "id": Node {
              "end": 1099,
              "loc": SourceLocation {
                "end": Position {
                  "column": 13,
                  "index": 1099,
                  "line": 51,
                },
                "filename": undefined,
                "identifierName": "logs",
                "start": Position {
                  "column": 9,
                  "index": 1095,
                  "line": 51,
                },
              },
              "name": "logs",
              "start": 1095,
              "type": "Identifier",
            },
            "kind": "resource",
            "resource": {
              "api": "<% \`logs/\${CTX.logId}\` %>",
              "config": {
                "enabled": "<% !!CTX.logId %>",
                "fallback": null,
              },
              "http": true,
              "name": "logs",
              "transform": "<% DATA.data.logs %>",
            },
          },
        },
        "children": [
          {
            "children": [
              {
                "children": [
                  {
                    "children": undefined,
                    "events": {
                      "search": [
                        {
                          "action": "conditional",
                          "payload": {
                            "alternate": [
                              {
                                "action": "update_variable",
                                "payload": {
                                  "name": "logId",
                                  "value": undefined,
                                },
                              },
                              {
                                "action": "update_query",
                                "payload": {
                                  "args": [
                                    {
                                      "logId": null,
                                    },
                                    {
                                      "notify": false,
                                    },
                                  ],
                                  "method": "push",
                                },
                              },
                            ],
                            "consequent": [
                              {
                                "action": "update_variable",
                                "payload": {
                                  "name": "logId",
                                  "value": "<% EVENT.detail %>",
                                },
                              },
                              {
                                "action": "update_query",
                                "payload": {
                                  "args": [
                                    {
                                      "logId": "<% EVENT.detail %>",
                                    },
                                    {
                                      "notify": false,
                                    },
                                  ],
                                  "method": "push",
                                },
                              },
                            ],
                            "test": "<% EVENT.detail %>",
                          },
                        },
                      ],
                    },
                    "name": "Search",
                    "properties": {
                      "inputStyle": {
                        "width": "300px",
                      },
                      "placeholder": "输入任务 ID 并按回车",
                      "value": "<%= CTX.logId %>",
                    },
                    "ref": undefined,
                    "reference": null,
                  },
                  {
                    "children": [
                      {
                        "children": [
                          {
                            "name": "Plaintext",
                            "properties": {
                              "textContent": "<%= CTX.allExpanded ? "全部收起" : "全部展开" %>",
                            },
                          },
                        ],
                        "events": {
                          "click": [
                            {
                              "action": "update_variable",
                              "payload": {
                                "name": "allExpanded",
                                "value": "<% !CTX.allExpanded %>",
                              },
                            },
                          ],
                        },
                        "name": "Button",
                        "properties": {},
                        "ref": undefined,
                        "reference": null,
                      },
                    ],
                    "name": "If",
                    "properties": {
                      "dataSource": "<%= CTX.logs %>",
                    },
                  },
                ],
                "events": undefined,
                "name": "Toolbar",
                "properties": {},
                "ref": undefined,
                "reference": null,
              },
              {
                "children": [
                  {
                    "children": undefined,
                    "events": undefined,
                    "name": "Table",
                    "properties": {
                      "columns": [
                        {
                          "dataIndex": "time",
                          "key": "time",
                          "render": {
                            "children": [
                              {
                                "name": "Plaintext",
                                "properties": {
                                  "textContent": "<%= FN._helper_mergeTexts({type:"expression",value:(moment(v * 1000).format("HH:mm:ss"))}) %>",
                                },
                              },
                            ],
                            "params": [
                              "v",
                            ],
                          },
                          "title": "时间",
                        },
                        {
                          "dataIndex": "opType",
                          "key": "opType",
                          "render": {
                            "children": [
                              {
                                "name": "Plaintext",
                                "properties": {
                                  "textContent": "<%= FN._helper_mergeTexts({type:"expression",value:(\`\${v}\${r.opSubType ? \` / \${r.opSubType}\` : ""}\`)}) %>",
                                },
                              },
                            ],
                            "params": [
                              "v",
                              "r",
                            ],
                          },
                          "title": "操作类型",
                        },
                        {
                          "dataIndex": "data",
                          "key": "data",
                          "render": {
                            "children": [
                              {
                                "children": [
                                  {
                                    "children": undefined,
                                    "events": undefined,
                                    "name": "CodeBlock",
                                    "properties": {
                                      "language": "json",
                                      "source": "<%= FN.prettierOutput(v) %>",
                                      "style": {
                                        "maxWidth": "696px",
                                      },
                                    },
                                    "ref": undefined,
                                    "reference": null,
                                  },
                                  {
                                    "children": [
                                      {
                                        "name": "Plaintext",
                                        "properties": {
                                          "textContent": "<%= v %>",
                                        },
                                      },
                                    ],
                                    "events": undefined,
                                    "name": "Plaintext",
                                    "properties": {
                                      "style": {
                                        "display": "block",
                                        "maxWidth": "696px",
                                        "overflow": "hidden",
                                        "textOverflow": "ellipsis",
                                        "whiteSpace": "nowrap",
                                      },
                                    },
                                    "ref": undefined,
                                    "reference": null,
                                    "slot": "else",
                                  },
                                ],
                                "name": "If",
                                "properties": {
                                  "dataSource": "<%= CTX.allExpanded || CTX.expandedSet.has(r.logId) %>",
                                },
                              },
                            ],
                            "params": [
                              "v",
                              "r",
                            ],
                          },
                          "title": "数据",
                        },
                        {
                          "dataIndex": "operation",
                          "key": "operation",
                          "render": {
                            "children": [
                              {
                                "children": undefined,
                                "events": {
                                  "open.detail": [
                                    {
                                      "action": "update_variable",
                                      "payload": {
                                        "name": "detailLog",
                                        "value": "<% r %>",
                                      },
                                    },
                                    {
                                      "action": "call_ref",
                                      "payload": {
                                        "args": [],
                                        "method": "open",
                                        "ref": "modalRef",
                                        "scope": "global",
                                      },
                                    },
                                  ],
                                  "toggle": [
                                    {
                                      "action": "update_variable",
                                      "payload": {
                                        "name": "expandedSet",
                                        "value": "<% new Set(
                        CTX.expandedSet.has(r.logId)
                          ? [...CTX.expandedSet].filter((id) => id !== r.logId)
                          : [...CTX.expandedSet, r.logId]
                      ) %>",
                                      },
                                    },
                                  ],
                                },
                                "name": "Operations",
                                "properties": {
                                  "allExpanded": "<%= CTX.allExpanded %>",
                                  "expanded": "<%= CTX.expandedSet.has(r.logId) %>",
                                },
                                "ref": undefined,
                                "reference": {
                                  "name": "Operations",
                                  "type": "local",
                                },
                              },
                            ],
                            "params": [
                              "v",
                              "r",
                            ],
                          },
                          "title": "操作",
                        },
                      ],
                      "dataSource": {
                        "list": "<%= CTX.logs %>",
                      },
                      "pagination": false,
                      "rowKey": "logId",
                    },
                    "ref": undefined,
                    "reference": null,
                  },
                ],
                "name": "If",
                "properties": {
                  "dataSource": "<%= CTX.logs %>",
                },
              },
              {
                "children": [
                  {
                    "children": [
                      {
                        "children": undefined,
                        "events": undefined,
                        "name": "Descriptions",
                        "properties": {
                          "columns": 1,
                          "dataSource": "<%= CTX.detailLog %>",
                          "list": [
                            {
                              "field": "logId",
                              "label": "日志 ID",
                            },
                            {
                              "field": "conversationId",
                              "label": "任务 ID",
                            },
                            {
                              "field": "taskId",
                              "label": "任务 ID",
                            },
                            {
                              "field": "time",
                              "label": "时间",
                              "render": {
                                "children": [
                                  {
                                    "name": "Plaintext",
                                    "properties": {
                                      "textContent": "<%= FN._helper_mergeTexts({type:"expression",value:(moment(r.time * 1000).format("llll"))}) %>",
                                    },
                                  },
                                ],
                                "params": [
                                  "r",
                                ],
                              },
                            },
                          ],
                        },
                        "ref": undefined,
                        "reference": null,
                      },
                      {
                        "children": undefined,
                        "events": undefined,
                        "name": "CodeBlock",
                        "properties": {
                          "language": "json",
                          "source": "<%= FN.prettierOutput(CTX.detailLog.data) %>",
                        },
                        "ref": undefined,
                        "reference": null,
                      },
                    ],
                    "name": "If",
                    "properties": {
                      "dataSource": "<%= CTX.detailLog %>",
                    },
                  },
                ],
                "events": undefined,
                "name": "Modal",
                "properties": {
                  "hideCancelButton": true,
                  "maskClosable": true,
                  "title": "<%= CTX.detailLog
            ? \`\${r.opType}\${r.opSubType ? \` / \${r.opSubType}\` : ""}\`
            : "" %>",
                  "width": "800px",
                },
                "ref": "modalRef",
                "reference": null,
              },
            ],
            "events": undefined,
            "name": "View",
            "properties": {
              "title": "任务日志",
            },
            "ref": undefined,
            "reference": null,
          },
        ],
        "id": null,
        "type": "view",
      },
      "title": "任务日志",
      "type": "view",
    },
    "errors": [],
    "filePath": "View.tsx",
    "internals": [
      {
        "component": {
          "bindingMap": Map {
            Node {
              "end": 199,
              "loc": SourceLocation {
                "end": Position {
                  "column": 10,
                  "index": 199,
                  "line": 14,
                },
                "filename": undefined,
                "identifierName": "expanded",
                "start": Position {
                  "column": 2,
                  "index": 191,
                  "line": 14,
                },
              },
              "name": "expanded",
              "range": undefined,
              "start": 191,
              "type": "Identifier",
            } => {
              "id": Node {
                "end": 199,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 10,
                    "index": 199,
                    "line": 14,
                  },
                  "filename": undefined,
                  "identifierName": "expanded",
                  "start": Position {
                    "column": 2,
                    "index": 191,
                    "line": 14,
                  },
                },
                "name": "expanded",
                "range": undefined,
                "start": 191,
                "type": "Identifier",
              },
              "initialValue": undefined,
              "kind": "param",
            },
            Node {
              "end": 214,
              "loc": SourceLocation {
                "end": Position {
                  "column": 13,
                  "index": 214,
                  "line": 15,
                },
                "filename": undefined,
                "identifierName": "allExpanded",
                "start": Position {
                  "column": 2,
                  "index": 203,
                  "line": 15,
                },
              },
              "name": "allExpanded",
              "range": undefined,
              "start": 203,
              "type": "Identifier",
            } => {
              "id": Node {
                "end": 214,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 13,
                    "index": 214,
                    "line": 15,
                  },
                  "filename": undefined,
                  "identifierName": "allExpanded",
                  "start": Position {
                    "column": 2,
                    "index": 203,
                    "line": 15,
                  },
                },
                "name": "allExpanded",
                "range": undefined,
                "start": 203,
                "type": "Identifier",
              },
              "initialValue": undefined,
              "kind": "param",
            },
            Node {
              "end": 226,
              "loc": SourceLocation {
                "end": Position {
                  "column": 10,
                  "index": 226,
                  "line": 16,
                },
                "filename": undefined,
                "identifierName": "onToggle",
                "start": Position {
                  "column": 2,
                  "index": 218,
                  "line": 16,
                },
              },
              "name": "onToggle",
              "range": undefined,
              "start": 218,
              "type": "Identifier",
            } => {
              "id": Node {
                "end": 226,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 10,
                    "index": 226,
                    "line": 16,
                  },
                  "filename": undefined,
                  "identifierName": "onToggle",
                  "start": Position {
                    "column": 2,
                    "index": 218,
                    "line": 16,
                  },
                },
                "name": "onToggle",
                "range": undefined,
                "start": 218,
                "type": "Identifier",
              },
              "kind": "eventHandler",
            },
            Node {
              "end": 242,
              "loc": SourceLocation {
                "end": Position {
                  "column": 14,
                  "index": 242,
                  "line": 17,
                },
                "filename": undefined,
                "identifierName": "onOpenDetail",
                "start": Position {
                  "column": 2,
                  "index": 230,
                  "line": 17,
                },
              },
              "name": "onOpenDetail",
              "range": undefined,
              "start": 230,
              "type": "Identifier",
            } => {
              "id": Node {
                "end": 242,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 14,
                    "index": 242,
                    "line": 17,
                  },
                  "filename": undefined,
                  "identifierName": "onOpenDetail",
                  "start": Position {
                    "column": 2,
                    "index": 230,
                    "line": 17,
                  },
                },
                "name": "onOpenDetail",
                "range": undefined,
                "start": 230,
                "type": "Identifier",
              },
              "kind": "eventHandler",
            },
          },
          "children": [
            {
              "children": [
                {
                  "children": [
                    {
                      "children": undefined,
                      "events": {
                        "click": [
                          {
                            "action": "dispatch_event",
                            "payload": {
                              "detail": "<% EVENT.detail %>",
                              "type": "toggle",
                            },
                          },
                        ],
                      },
                      "name": "Button",
                      "properties": {
                        "textContent": "<%= FN._helper_mergeTexts({type:"expression",value:(STATE.expanded ? "收起" : "展开")}) %>",
                      },
                      "ref": undefined,
                      "reference": null,
                    },
                  ],
                  "name": "If",
                  "properties": {
                    "dataSource": "<%= !STATE.allExpanded %>",
                  },
                },
                {
                  "children": undefined,
                  "events": {
                    "click": [
                      {
                        "action": "dispatch_event",
                        "payload": {
                          "detail": "<% EVENT.detail %>",
                          "type": "open.detail",
                        },
                      },
                    ],
                  },
                  "name": "Button",
                  "properties": {
                    "textContent": "详情",
                  },
                  "ref": undefined,
                  "reference": null,
                },
              ],
              "events": undefined,
              "name": "Toolbar",
              "properties": {},
              "ref": undefined,
              "reference": null,
            },
          ],
          "id": Node {
            "end": 186,
            "loc": SourceLocation {
              "end": Position {
                "column": 19,
                "index": 186,
                "line": 13,
              },
              "filename": undefined,
              "identifierName": "Operations",
              "start": Position {
                "column": 9,
                "index": 176,
                "line": 13,
              },
            },
            "name": "Operations",
            "start": 176,
            "type": "Identifier",
          },
          "type": "template",
        },
        "type": "template",
      },
      {
        "function": {
          "name": "prettierOutput",
          "source": "function prettierOutput(output: string) {
  try {
    const data = JSON.parse(output);
    return JSON.stringify(data, null, 2);
  } catch (e) {
    return output;
  }
}",
          "typescript": true,
        },
        "type": "function",
      },
    ],
    "moduleType": "entry",
    "namedExports": Map {},
    "source": undefined,
    "usedHelpers": Set {
      "_helper_mergeTexts",
    },
  },
  "errors": [],
  "files": undefined,
  "modules": [],
}
`;
